airflow/providers/sftp/hooks/sftp.py

```diff
@@ -400,3 +406,177 @@ def get_files_by_pattern(self, path, fnmatch_pattern) -> list[str]:
+            "host": conn.host,
+            "port": conn.port,
+            "username": conn.login,
+            "password": conn.password,
+        }
+        if self.key_file:
+            conn_config.update(client_keys=self.key_file)
+        if self.known_hosts:
+            if self.known_hosts.lower() == "none":
+                conn_config.update(known_hosts=None)
+            else:
+                conn_config.update(known_hosts=self.known_hosts)
+        if self.private_key:
+            _private_key = asyncssh.import_private_key(self.private_key, self.passphrase)
+            conn_config.update(client_keys=[_private_key])
+        if self.passphrase:
+            conn_config.update(passphrase=self.passphrase)
+        ssh_client_conn = await asyncssh.connect(**conn_config)
+        return ssh_client_conn
+
+    async def list_directory(self, path: str = "") -> list[str] | None:
+        """Returns a list of files on the SFTP server at the provided path."""
+        ssh_conn = await self._get_conn()
+        sftp_client = await ssh_conn.start_sftp_client()
+        try:
+            files = await sftp_client.listdir(path)
+            return sorted(files)
+        except asyncssh.SFTPNoSuchFile:
+            return None
+
+    async def read_directory(self, path: str = "") -> Sequence[asyncssh.sftp.SFTPName] | None:
+        """Returns a list of files along with their attributes on the SFTP server at the provided path."""
+        ssh_conn = await self._get_conn()
+        sftp_client = await ssh_conn.start_sftp_client()
+        try:
+            files = await sftp_client.readdir(path)
+            return files
+        except asyncssh.SFTPNoSuchFile:
+            return None
+
+    async def get_files_and_attrs_by_pattern(
+        self, path: str = "", fnmatch_pattern: str = ""
+    ) -> Sequence[asyncssh.sftp.SFTPName]:
+        """
+        Get the files along with their attributes matching the pattern (e.g. ``*.pdf``) at the provided path.
+
+        if one exists. Otherwise, raises an AirflowException to be handled upstream for deferring
+        """
+        files_list = await self.read_directory(path)
+        if files_list is None:
+            raise FileNotFoundError(f"No files at path {path!r} found...")
+        matched_files = [file for file in files_list if fnmatch(str(file.filename), fnmatch_pattern)]
+        return matched_files
+
+    async def get_files_by_pattern(self, path: str = "", fnmatch_pattern: str = "") -> list[str]:
+        """
+        Returns the name of a file matching the file pattern at the provided path.
+
+        If one exists Otherwise, raises an AirflowException to be handled upstream for deferring.
+        """
+        files_list = await self.list_directory(path)
+        if files_list is None:
+            raise AirflowException(f"No files at path {path} found...")
+        matched_files = [file for file in files_list if fnmatch(file, fnmatch_pattern)]
+        return matched_files
```

Comment:
I think we can remove this method as it's not getting used here and also cleanup the unit tests for this method.

Reply:
+1

Reply:
Make sense, removed it.